### PR TITLE
Meta-mint SLD

### PR DESCRIPTION
**Pull request by aquiladev**

### CODE REVIEW

File: contracts/util/WhitelistedMinter.sol [hunk 1]
```diff
@@ -157,4 +166,41 @@ contract WhitelistedMinter is IMintingController, BulkWhitelistedRole {
...
+            .recover(signature);
+        require(signer != address(0), 'WhitelistedMinter: SIGNATURE_IS_INVALID');
+        require(isWhitelisted(signer), 'WhitelistedMinter: SIGNER_IS_NOT_WHITELISTED');
+    }
+
+    /**
+     * Proxy is an alternative solution for meta-transactions
+     * Disadvantages:
+     *  - the function can proxy any call, even we don't need to do this for some functions
+     *  - in order to execute the logic the contract should be whitelisted by its own. lol
```
Comment (bogdan) [inline contracts/util/WhitelistedMinter.sol:182]: Can this be negated by extending `isWhitelistedMinter` with `_sender == address(this)` or by adding `_addWhitelisted(address(this)` in constructor?

File: contracts/util/WhitelistedMinter.sol [hunk 2]
```diff
@@ -64,6 +67,12 @@ contract WhitelistedMinter is IMintingController, BulkWhitelistedRole {
         _mintingController.mintSLD(to, label);
     }
 
+    function mintSLDFor(address to, string calldata label, bytes calldata signature) external {
```
Comment (bogdan) [inline contracts/util/WhitelistedMinter.sol:70]: Do you think it is fine that we will have meta-tx version only for function wihtout resolver and preconfigure?

File: contracts/util/WhitelistedMinter.sol [hunk 3]
```diff
@@ -64,6 +67,12 @@ contract WhitelistedMinter is IMintingController, BulkWhitelistedRole {
...
+        bytes4 selector = bytes4(keccak256('mintSLD(address,string)'));
```
Comment (bogdan) [inline contracts/util/WhitelistedMinter.sol:71]: Is it expensive to calculate on each call?

File: contracts/util/WhitelistedMinter.sol [hunk 4]
```diff
@@ -145,6 +194,31 @@ contract WhitelistedMinter is IMintingController, BulkWhitelistedRole {
...
+        (bool success, bytes memory result) = address(this).call(data);
+        if (success == false) {
+            /* solium-disable-next-line security/no-inline-assembly */
+            assembly {
+                let ptr := mload(0x40)
+                let size := returndatasize
+                returndatacopy(ptr, 0, size)
+                revert(ptr, size)
+            }
+        }
```
Comment (aquiladev) [inline contracts/util/WhitelistedMinter.sol:218]: @bogdan what do you think about event's emission on success call?

**Review by bogdan — APPROVED** [review]
Add an event and looks good.

File: contracts/util/WhitelistedMinter.sol [hunk 5]
```diff
@@ -15,6 +15,8 @@ import "../Resolver.sol";
 contract WhitelistedMinter is IMintingController, BulkWhitelistedRole {
     using ECDSA for bytes32;
 
+    event Relayed(address indexed sender, bytes32 dataHash);
```
Comment (bogdan) [inline contracts/util/WhitelistedMinter.sol:18]: Is it easy to add an indexed field by the function hash being called? It can allow easier debugging?

**Pull request merged**